data_handler.py

```diff
@@ -1,37 +1,61 @@
+# title string, length int (sec), link string, filepath string, thumbnail string
 def insert_data(title, length, link, filepath, thumbnail=""):
-    if length > 60 * MAX_LENGTH:
-        return "Too long to be stored."
-    timestamp = datetime.datetime.now()
-    with sqlite3.connect(info_db) as c:
-        c.execute("""CREATE TABLE IF NOT EXISTS info_db (time_ timestamp, title text, length real, link text, filepath text, thumbnail text);""")
-        c.execute('''INSERT into info_db VALUES (?,?,?,?,?);''',(timestamp,title,length,link, filepath, thumbnail))
-    return "Data inserted."
+    if length > 60 * MAX_LENGTH or link[:20]!="youtube.com/watch?v=":
+        return False
+    if not retrieve_data(link):
+        timestamp = datetime.datetime.now()
+        with sqlite3.connect(info_db) as c:
+            c.execute("""CREATE TABLE IF NOT EXISTS info_db (time_ timestamp, title text, length real, link text, filepath text, thumbnail text);""")
+            c.execute('''INSERT into info_db VALUES (?,?,?,?,?,?);''',(timestamp,title,length,link, filepath, thumbnail))
+    return True
 
+# given link: return row data, if exists
 def retrieve_data(link):
-    pass
+    with sqlite3.connect(info_db) as c:
+        c.execute("""CREATE TABLE IF NOT EXISTS info_db (time_ timestamp, title text, length real, link text, filepath text, thumnbnail text);""")
+        current = c.execute('''SELECT * FROM info_db WHERE link=;''',(link)).fetchall()
+    return current
+
+# given link: deletes row data, if exists. returns LIST of mp3 filepath of deleted entries.
+def delete_data(link):
+    with sqlite3.connect(info_db) as c:
+        c.execute("""CREATE TABLE IF NOT EXISTS info_db (time_ timestamp, title text, length real, link text, filepath text, thumnbnail text);""")
+        out = c.execute('''SELECT filepath FROM info_db WHERE link=?;''',(link)).fetchall()
+        c.execute('''DELETE * FROM info_db WHERE link=?;''',(link))
+        return out
 
+# clear data entries older than timestamp given, if exists. returns LIST of mp3 filepaths of deleted entries.
 def remove_old_data(timestamp):
     with sqlite3.connect(info_db) as c:
         c.execute("""CREATE TABLE IF NOT EXISTS info_db (time_ timestamp, title text, length real, link text, filepath text, thumnbnail text);""")
-        c.execute('''DELETE * FROM ht_db WHERE time_>=?;''',(timestamp))
-    return "Old data deleted."
+        out = c.execute('''SELECT filepath FROM info_db WHERE time_<=?;''',(timestamp)).fetchall()
+        c.execute('''DELETE * FROM info_db WHERE time_<=?;''',(timestamp))
+        return out
+
+# very silly & inefficient for large tables. returns size of table.
+def check_size():
+    with sqlite3.connect(info_db) as c:
+        c.execute("""CREATE TABLE IF NOT EXISTS info_db (time_ timestamp, title text, length real, link text, filepath text, thumnbnail text);""")
+        return len(c.execute('''SELECT length FROM info_db;''').fetchall())
```

Comment:
Think there's a quick way to get size of table in sql, but yea doesn't matter rn

Reply:
Yea think this code will work fine since all rows are unique:
`c.execute("""SELECT COUNT(*) FROM info_db;""").fetchone()`